pkg/api/auth/public_api_handler.go

```diff
@@ -112,6 +114,45 @@ func (h *PublicAPIHandler) login(input *Login) (tokenResp *TokenResponse, status
 		}
 	}()
 
+	var impersonateAuthInfo *clientcmdapi.AuthInfo
+	if input.Username != "" && input.Password != "" && input.Provider != "" {
+		impersonateAuthInfo, status, err = h.userlogin(input)
+		if err != nil {
+			return nil, status, err
+		}
+	} else if input.Token != "" || input.KubeConfig != "" {
+		impersonateAuthInfo, status, err = h.k8slogin(input)
+		if err != nil {
+			return nil, status, err
+		}
+	} else {
+		return nil, http.StatusBadRequest, errors.New("empty authentication information")
+	}
+
+	var token string
+	token, err = h.tokenManager.Generate(*impersonateAuthInfo)
+	if err != nil {
+		return nil, http.StatusInternalServerError, errors.Wrapf(err, "Failed to generate token")
+	}
+
+	escapedToken := url.QueryEscape(token)
+	return &v1alpha1.TokenResponse{JWEToken: escapedToken}, http.StatusOK, nil
+}
+
+func (h *PublicAPIHandler) userlogin(input *v1alpha1.Login) (*clientcmdapi.AuthInfo, int, error) {
+	provider, err := providers.GetProvider(input.Provider)
+	if err != nil {
+		return nil, http.StatusBadRequest, errors.Wrapf(err, "Invalid provider")
+	}
+
+	userName, err := provider.AuthenticateUser(input)
+	if err != nil {
+		return nil, http.StatusUnauthorized, errors.Wrapf(err, "Failed to login")
+	}
+	return &clientcmdapi.AuthInfo{Impersonate: userName}, http.StatusOK, nil
+}
+
+func (h *PublicAPIHandler) k8slogin(input *v1alpha1.Login) (*clientcmdapi.AuthInfo, int, error) {
```

Comment:
If you are keeping k8s login logic, should it be part of providers too?

Reply:
remove provider